# CHRISTOPHER IAN HUNTER

**PERSONAL DETAILS:**

**Home Address:** 3B, Tower 5, Deerhill Bay **Date of Birth :** 9th May 1975

4699 Tai Po Rd, **Tel** : +(852) 5593 7602

Tai Po, N.T. Hong Kong. **email:** [only1chunts@hotmail.com](mailto:only1chunts@hotmail.com)

## EMPLOYMENT HISTORY:

**March 2013 to present:** BGI-Hong Kong Co Ltd. 16 Dai Fu Street, Tai Po Industrial estate, Tai Po, N.T. Hong Kong.
• Lead BioCurator, GigaDB. GigaScience is an online, open-access journal that includes, as part of its publishing activities, the database _GigaDB_. GigaScience recognised the need for the association of the data with the manuscripts and was the first to provide the dedicated resources to achieve this. _GigaDB_ primarily serves as a repository to host data and tools associated with articles in GigaScience. Through our association with DataCite, each dataset in GigaDB is assigned a DOI that can be used as a standard citation for future use of these data in other articles by the authors and other researchers My role within the team is ; to assist users with the organisation of their data, and making sure all the required files are deposited, both in GigaDB and external resources (if required); to liaise with users and potential users to prepare GigaDB for future expansion; to conceptualise and implement appropriate search and visualisation tools to enable users to make full use of the data stored in GigaDB. The scope of scientific data we handle is large and requires a great deal of open collaboration with many people and groups around the world.

**2008 to Present:** Voluntary involvement in the Genomic Standards Consortium.
• As an active member of the working groups involved in term vetting and tool developments, my daily experiences with the data that is being standardised is invaluable to the generation of both comprehensive and workable standards for the (meta)genome sequencing community.

**Sept 2009 to Sept 2012:** EMBL-EBI, Wellcome Trust Genome Campus, Cambridge, CB10 1SD.
• Bioinformatician, InterPro-Metagenomics. The EBI metagenomics portal is a resource providing access to the researchers in the relatively new field of metagenomics. It has the aim of being a user friendly interface to some currently available EBI resources (e.g. InterPro, ENA, UniProt/MES, GO, Ensembl), promoting their utility in the field of metagenomics. I have a lead roll in the progression of this portal with input into all aspects from web design, usability, software implementation and data acquisition from users. The role requires a great deal of time management and communication skills. I present the progress at regular monthly updates with team leaders and have been to various international conferences to promote the EBIs involvement in this field.

**Sept 2007 to Sept 2009: EMBL-EBI, Wellcome Trust Genome Campus, Cambridge, CB10 1SD.**
•  Database curator, EMBL-bank nucleotide sequence database.
EMBL-bank constitutes Europe&#39;s primary nucleotide sequence resource. DNA and RNA sequences are submitted from individual researchers, genome sequencing projects and patent applications. EMBL-bank provides an up to date searchable database accessible to all and my role within the team is as a curator.This role entails overseeing entry of submissions to ensure the database remains accurate and searchable. The tools to undertake this process are mainly Unix based, with the powerful word-processor Emacs being heavily relied upon. As a curator I am the primary contact point for most submitters and as such it is essential that I use all my communication skills to ensure the &quot;customer&quot; feels in control of their data and yet still acquire sufficient information about each submission to fulfil the requirements of a minimum standard. I am also taking a leading role in the development of submission protocols for the reposition of the new sequence technology data (short read archive). Participation in bi-weekly meetings with other curators helps ensure a uniform standard for all members of the team.

**Feb 2004 to Sept 2007: Sanger Institute, Wellcome Trust Genome Campus, Cambridge, CB10 1SA.**
• Postdoctoral Researcher, Cancer Genome Project. The aim of the project is to determine somatic variants within the DNA taken from various cancer tissues to help further the understanding of how cancer occurs. My role within the team was to oversee the entire process from primer design for the PCR and sequencing, through to the analysis of the sequence data. Various software applications were used including Mutation Surveyor, Chromas and OncoCSA. Due to the large datasets a lot of sample tracking was required, initially this was done in MS Excel, so became very proficient with its use. However the in house IT department developed an integrated sequence analysis and laboratory tracking system. I was involved in the development process at the user level, which required the full use of my communication skills to help the IT staff understand the biological nature of what the software needed to do. I also presented updates and potentially interesting results in weekly meetings with other senior members of the project.

**July 2002 to Jan 2004: MRCUK HGMP Resource Centre, Wellcome Trust Genome Campus, Cambridge CB10 1SB**
• Postdoctoral Researcher. 
I recognised problems and implemented rectifications, to establish a consistent and useful Dual Luciferase assay, which the group uses to help identify regulatory regions within the 3&#39;UTR of genes. I controlled the flow of work performed by myself and my RA, through a self implemented web-based tracking system. Other key duties included target identification from literature and computational analysis, identification of suitable primers and cDNA clones, and training of my RA in all relevant techniques. Unfortunately I had to leave this role due to the funding body (RiboTargets) withdrawing its funding for commercial reasons.

**Jan 2002 to June 2002: Cylcacel Ltd. Polgen Division.**
• Scientific researcher. 
Work included; DNA and RNA preparation, PCR, gene cloning, transient transfections, FACS analysis and dsRNA synthesis. However the largest proportion of my time was spent augmenting automation of the in-house procedures primarily by the use of an MWG liquid handling robot.

**1998 to 2002: PhD.** 
• **Title:** Identification, cloning and characterisation of the Alpha2 Adrenergic Receptor gene family in the Japanese pufferfish** _ **Fugu rubripes** _ **
• **Submitted:** 10th May 2002. Darwin College, Cambridge University, Cambridge, UK. **Studies undertaken:** at MRC UK HGMP Resource Centre, in the Fugu Genomics Group, and at Pfizer Global R&amp;D, Sandwich, Kent, UK. **Funding Body:** BBSRC Industrial case award with Pfizer Global R&amp;D, Kent UK.

**Key Areas of Study:**
- Sequencing, cloning, PCR, hybridisation, RT-PCR, Genetics/genomics
- Evolutionary Phylogenetic analysis (Treetool, phylip and phylowin. )
- Pharmacological characterisation was partially undertaken by the use of FLIPR at Pfizer, Kent, UK.
- Computing software used during this project, in addition to those mentioned above, on Unix include; GCG seqlab, Staden Gap4 and Pregap4, Readseq, NIX, BLAST, SeaView and Clustal, and on Windows; Word, Excel, EndNote, BioEdit, and Powerpoint.

**1994 to 1998: BSc (Hons) Applied Biology 2:i**
The Nottingham Trent University, Nottingham, UK.

## ADDITIONAL SKILLS :
- **Computing skills;** Extensive use of various software including MS Office, excellent understanding of Unix commands and environment, use of a broad spectrum of web based utilities including SRS and Ensembl, limited programming/scripting ability in python, perl and R. Working knowledge of ontologies (Protégé), XML (Oxygen) and SQL.
- **Personal skills;** Ability to lead effectively, confident and competent in communication both verbal and written.

## COURSES, CONFERENCES AND PRESENTATIONS
- Various &quot;Introduction to&quot; course including; Management, Basic Unix, Staden Phylogeny Scientific writing course, presenting with impact.
- Regular oral presentations to update the group on current work.
- Presentation to secondary school students and lay audiences about genetics and cancer
- Introduction to Perl programming, 3 days, covered basic theory and practice.
- Introduction to Python programming, basic theory
- Hands-on Introduction to XML
- Various Metagenomic conferences
- Various Next generation sequences course and conferences including: New sequencing technologies workshop 08, 09 and 2010.,
- Genomic Standards Consortium meetings; GSC6, 9, 11, 12, 14 and 15.
- Teaching/demonstrating EBI Metagenomics resources at EU-US training course for MicroB3.

## ACTIVITIES AND INTERESTS

I enjoy numerous sporting activities including badminton, squash, swimming and volleyball. In the past I have represented and captained my sixth form rugby team, as well as playing for a local club and the county, unfortunately due to injury I no longer play, but have recently taken up a coaching role for the local under 9&#39;s team. Additionally I enjoy many other outdoor activities, such as hiking, jogging and cycling. Other interests include gardening and DIY.

## PUBLICATIONS

Hunter S, …, **Hunter C** , ..., Sansone SA. (2013) &quot;EBI metagenomics--a new resource for the analysis and archiving of metagenomic data.&quot; Nucleic acids research, 2013 Oct 27. [Epub ahead of print]

Stoev,P. ... **Hunter,C.I.** , … Penev, L. &quot;Eupolybothrus cavernicolus Komerički &amp; Stoev sp.n (Chilopoda: Lithobiomorpha: Lithobiidae): the first eukaryotic species description combining transcriptomic, DNA barcoding and micro-CT imaging data expand article info.&quot; Biodiversity Data Journal 1: e1013 (28 Oct 2013) doi:10.3897/BDJ.1.e1013

**Hunter, C.** I., Mitchell, A., Jones, P., McAnulla C., Pesseat, S., Scheremetjew M. and Hunter, S. (2012) &quot;Metagenomic analysis; the challenge of the data bonanza.&quot; Briefings in Bioinformatics 2012; 13(6): 743-746

**Hunter, C**., Cochrane, G., Apweiler, R. and Hunter, S. (2011) The EBI Metagenomics Archive, Integration and Analysis Resource, in Handbook of Molecular Microbial Ecology I: Metagenomics and Complementary Approaches (ed F. J. de Bruijn), John Wiley &amp; Sons.

Amid C, Birney E, … **Hunter C,** … Cochrane G. &quot;Major submission tool developments at the European nucleotide archive.&quot; Nucleic Acids Res. 2012 Jan; 40(Database issue):D43-47.

Yilmaz P, Kottmann R,…, **Hunter C** , …, Glöckner FO. &quot;Minimum information about a marker gene sequence (MIMARKS) and minimum information about any (x) sequence (MIxS) specifications.&quot; Nat Biotechnol. 2011 May;29(5):415-20.

Leinonen R, Akhtar R, Birney E, …, **Hunter C** , …, Cochrane G. &quot;Improvements to services at the European Nucleotide Archive.&quot; Nucleic Acids Res. 2010 Jan;38(Database issue):D39-45.

Cochrane G, Akhtar R, …, **Hunter C** , …, Birney E. &quot;Petabyte-scale innovations at the European Nucleotide Archive.&quot; Nucleic Acids Res. 2009 Jan;37(Database issue):D19-25.

Greenman C, Stephens P, Smith R, Dalgliesh GL, **Hunter C** , Wooster R, Futreal PA, Stratton MR, _et al_.

&quot;Patterns of somatic mutation in human cancer genomes.&quot;

Nature. 2007 Mar 8;446(7132):153-8.

Bignell G, Smith R, **Hunter C** , Futreal PA, Wooster R, Stratton MR, _et al_.

&quot;Sequence analysis of the protein kinase gene family in human testicular germ-cell tumors of adolescents and adults.&quot; Genes Chromosomes Cancer. 2006 Jan;45(1):42-6.

**Hunter C** , Smith R, Cahill DP, Stratton MR, Futreal AP, Wooster R, _et al_.

&quot;A hypermutation phenotype and somatic MSH6 mutations in recurrent human

malignant gliomas after alkylator chemotherapy&quot;. Cancer Research. 2006 April; 66:3987-91.

Stephens P, Edkins S, **Hunter C** , Stratton MR, Futreal PA, Wooster R, _et al_.

&quot;A screen of the complete protein kinase gene family identifies diverse patterns of somatic mutations in human breast cancer.&quot; Nat Genet. 2005 Jun;37(6):590-2.

Davies H, **Hunter C** , Smith R, Wooster R, Stratton MR, Futreal PA, _et al_.

&quot;Somatic mutations of the protein kinase gene family in human lung cancer.&quot;

Cancer Res. 2005 Sep 1;65(17):7591-5.

Stephens P, **Hunter C** , Bignell G, Futreal PA, Wooster R, Stratton MR, _et al_.

&quot;Lung cancer: intragenic ERBB2 kinase mutations in tumours.&quot;

Nature. 2004 Sep 30;431(7008):525-6.
